source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,198 @@
+  *command_line_options.mutable_duration() = std::move(duration);
+
+  ENVOY_LOG_MISC(info, "Sending load: {}", command_line_options.DebugString());
+  absl::StatusOr<nighthawk::client::ExecutionResponse> nighthawk_response_or =
+      nighthawk_service_client_.PerformNighthawkBenchmark(nighthawk_service_stub,
+                                                          command_line_options);
+  if (!nighthawk_response_or.ok()) {
+    ENVOY_LOG_MISC(error, "Nighthawk Service error: {}: {}", nighthawk_response_or.status().code(),
+                   nighthawk_response_or.status().message());
+    return nighthawk_response_or.status();
+  }
+  nighthawk::client::ExecutionResponse nighthawk_response = nighthawk_response_or.value();
+
+  absl::StatusOr<BenchmarkResult> benchmark_result_or =
+      metrics_evaluator_.AnalyzeNighthawkBenchmark(nighthawk_response, spec,
+                                                   name_to_custom_plugin_map);
+  if (!benchmark_result_or.ok()) {
+    ENVOY_LOG_MISC(error, "Benchmark scoring error: {}: {}", benchmark_result_or.status().code(),
+                   benchmark_result_or.status().message());
+    return benchmark_result_or.status();
+  }
+  BenchmarkResult benchmark_result = benchmark_result_or.value();
+  for (const MetricEvaluation& evaluation : benchmark_result.metric_evaluations()) {
+    ENVOY_LOG_MISC(info, "Evaluation: {}", evaluation.DebugString());
+  }
+  return benchmark_result;
+}
+
+absl::StatusOr<AdaptiveLoadSessionOutput> AdaptiveLoadControllerImpl::PerformAdaptiveLoadSession(
+    nighthawk::client::NighthawkService::StubInterface* nighthawk_service_stub,
+    const AdaptiveLoadSessionSpec& input_spec) {
+  AdaptiveLoadSessionSpec spec = session_spec_proto_helper_.SetSessionSpecDefaults(input_spec);
+  absl::Status validation_status = session_spec_proto_helper_.CheckSessionSpec(spec);
+  if (!validation_status.ok()) {
+    return validation_status;
+  }
+  absl::flat_hash_map<std::string, MetricsPluginPtr> name_to_custom_metrics_plugin_map =
+      LoadMetricsPlugins(spec);
+  StepControllerPtr step_controller = LoadStepControllerPluginFromSpec(spec);
+  AdaptiveLoadSessionOutput output;
+
+  // Threshold specs are reproduced in the output proto for convenience.
+  for (const nighthawk::adaptive_load::MetricSpecWithThreshold& threshold :
+       spec.metric_thresholds()) {
+    *output.mutable_metric_thresholds()->Add() = threshold;
+  }
+
+  Envoy::MonotonicTime start_time = time_source_.monotonicTime();
+  std::string doom_reason;
+  do {
+    absl::StatusOr<BenchmarkResult> result_or = PerformAndAnalyzeNighthawkBenchmark(
+        nighthawk_service_stub, spec, name_to_custom_metrics_plugin_map, *step_controller,
+        spec.measuring_period());
+    if (!result_or.ok()) {
+      return result_or.status();
+    }
+    BenchmarkResult result = result_or.value();
+    *output.mutable_adjusting_stage_results()->Add() = result;
+    step_controller->UpdateAndRecompute(result);
```

Comment:
It took me a second to follow the logic because step_controller->GetCurrentCommandLineOptions and step_controller->UpdateAndRecompute(result); are called in 2 different methods. 
I think it would be easier to follow if you moved step_controller->GetCurrentCommandLineOptions to this method.

Reply:
Just moved UpdateAndRecompute into the helper, which hopefully makes the same amount of sense. I had moved GetCurrentCommandLineOptions to the helper to reduce duplication in the main loop.